### PR TITLE
GH-4886 problem with use of @TestFactory and @TempDir

### PR DESCRIPTION
**Pull request by hmottestad**
GitHub issue resolved: #4886 

Briefly describe the changes proposed in this PR:

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/CONTRIBUTING.md) for more details):

 - [ ] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [ ] I've added tests for the changes I made
 - [ ] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [ ] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [ ] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

### CODE REVIEW

File: testsuites/sparql/src/main/java/org/eclipse/rdf4j/testsuite/query/parser/sparql/manifest/SPARQLComplianceTest.java [hunk 1]
```diff
@@ -67,18 +67,14 @@ public DynamicSparqlComplianceTest(String displayName, String testURI, String na
...
 				runTest();
-			} catch (Exception e) {
-				try {
-					tearDown();
-				} catch (Exception e2) {
-
-				}
+			} finally {
+				tearDown();
 			}
```
Comment (hmottestad) [inline testsuites/sparql/src/main/java/org/eclipse/rdf4j/testsuite/query/parser/sparql/manifest/SPARQLComplianceTest.java:78]: This code was swallowing exceptions and also the tearDown() method was only called if there was an exception.

**Comment by hmottestad** [comment]
@JervenBolleman I've run into a more general issue with your rewrite of the `SPARQLQueryComplianceTest` class et al. because the JUnit 5 `@TestFactory` annotation doesn't work very well with the `@TempDir` annotation. In fact `@TestFactory` doesn't work well with any of the lifecycle annotations, including before each. Essentially the combination of `@TestFactory` and `@TempDir` means that we use the same temporary directory for all the tests in a class, instead of a new sail and repository with a new temporary directory for each and every test.

This issue has been masked by the test code catching exceptions. Pretty much none of the NativeStore tests have been running. I managed to get the NativeStore tests working now, but the LmdbStore doesn't seem so happy about being continuously re-instantiated on top of the directory for an existing LmdbStore. @kenwenzel maybe you could take a look at why the LmdbStore gets upset after a while when being shutdown and then recreated on the same directory. You should be able to reproduce the issues by running `LmdbPARQL11UpdateComplianceTest`.

**Comment by JervenBolleman** [comment]
Oh no, I am so sorry for this!

**Comment by hmottestad** [comment]
"Aldri så galt at det ikke er godt for noe" - Norwegian way to say that even when bad things happen they still bring something good. The ExtensibleStore test failing led to discovering the issue with the tests here, and that led to discovering the issue with the LmdbStore :)

Maybe a quick fix is to add a File or Path param to the creeateRepo method and manually handle the creating and deletion of temporary directories.

Another approach would be to try to use the ParamaterizedTest annotation with the method source option. I do this with the ShaclSail where I read the directory structure to generate the tests at runtime.

It's still really nice to be able to run just a single test from IntelliJ though! Instead of always having to run every single one like I had to before.

**Comment by JervenBolleman** [comment]
I tried to create a dir per dynamic test. Still have the failing extensiblestore test. Native and Lmdb tests seem to run?
The code is in [pull:4888](https://github.com/eclipse-rdf4j/rdf4j/pull/4888)

**Comment by hmottestad** [comment]
Yeah, I think the ExtensibleStore failure is unrelated. It uses a ConcurrentHashMap as the main data structure when running the tests, so maybe there are some changes to that class in jdk 21 that is causing the issue. Not sure what it could be, and it's particularly annoying that I can't reproduce it in IntelliJ.

**Comment by kenwenzel** [comment]
@hmottestad I could imagine that something is not correctly closed when unsing the LmdbStore. Have you checked that everything gets closed after running the tests?

**Comment by hmottestad** [comment]
From what I could see everything is closed correctly, but I haven't done anything more than look over the code.

**Comment by kenwenzel** [comment]
Is the store instantiated with an empty directory or are there any artifacts left from the previous run?
It might also be the case that the store is not correctly closed due to some issue like
https://github.com/eclipse-rdf4j/rdf4j/issues/4806

I'll try to take a look within the next days.

**Comment by hmottestad** [comment]
> Is the store instantiated with an empty directory or are there any artifacts left from the previous run?
> 
> It might also be the case that the store is not correctly closed due to some issue like
> 
> https://github.com/eclipse-rdf4j/rdf4j/issues/4806
> 
> 
> 
> I'll try to take a look within the next days.

You can checkout this commit here: 5d825520e1c268222d6cb0163724c26ffa94f539

The issues occurs when the store is shutdown() and then recreated with the same directory. It doesn't happen immediately but after ~6 times.

**Comment by kenwenzel** [comment]
> > Is the store instantiated with an empty directory or are there any artifacts left from the previous run?
> > It might also be the case that the store is not correctly closed due to some issue like
> > #4806
> > I'll try to take a look within the next days.
> 
> You can checkout this commit here: [5d82552](https://github.com/eclipse-rdf4j/rdf4j/commit/5d825520e1c268222d6cb0163724c26ffa94f539)
> 
> The issues occurs when the store is shutdown() and then recreated with the same directory. It doesn't happen immediately but after ~6 times.

@hmottestad I nailed this down to an issue with the LMDB ValueStore and its recycling of IDs. I hope that I can fix this within the next days.

**Pull request merged**